File: benchmark/quadruped.py
# coding: utf8
import sys
import time

import crocoddyl
import numpy as np

from quadruped_walkgen import ActionModelQuadruped

N = 16  # number of nodes
T = int(sys.argv[1]) if (len(sys.argv) > 1) else int(5000)  # number of trials
MAXITER = (int(sys.argv[2]) if (len(sys.argv) > 1) else int(1))  # number of maximum iteration

# Gait Matrix
gait = np.zeros((6, 5))
fsteps = np.full((6, 13), np.nan)

# MpcInterface object that contains information about the current state of the robot
# Change the initial conditions here
lC = np.array([[0.0, 0.0, 0.2]]).transpose()  # CoM centered and at 20 cm above the ground
abg = np.array([[0.0, 0.0, 0.0]]).transpose()  # horizontal base (roll, pitch, 0.0)
lV = np.array([[0.2, 0.0, 0.0]]).transpose()  # motionless base (linear velocity)
lW = np.array([[0.0, 0.0, 0.0]]).transpose()  # motionless base (angular velocity)
l_feet = np.array([
    [0.19, 0.19, -0.19, -0.19],
    [0.15005, -0.15005, 0.15005, -0.15005],
    [0.0, 0.0, 0.0, 0.0],
])  # position of feet in local frame

x0 = np.vstack((lC, abg, lV, lW))  # Current state vector
# The reference state, copy of the initial position,
xref = np.repeat(x0, N + 1, axis=1)  # Desired future state vectors
xref[6, :] = 0.0  # Target linear velocity to Vx = 0

fsteps = np.array([
    [
        1.00000000e00,
        1.90776486e-01,
        1.48962816e-01,
        4.22498932e-03,
        1.90060159e-01,
        -1.50265109e-01,
        0.00000000e00,
        -1.89740429e-01,
        1.50467686e-01,
        1.78713224e-06,
        -1.90692335e-01,
        -1.48946056e-01,
        4.22561856e-03,
    ],
    [
        7.00000000e00,
        1.90776486e-01,
        1.48962816e-01,
        4.22498932e-03,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        -1.90692335e-01,
        -1.48946056e-01,
        4.22561856e-03,
    ],
    [
        1.00000000e00,
        1.90776486e-01,
        1.48962816e-01,
        4.22498932e-03,
        1.90000000e-01,
        -1.50050000e-01,
        0.00000000e00,
        -1.90000000e-01,
        1.50050000e-01,
        0.00000000e00,
        -1.90692335e-01,
        -1.48946056e-01,
        4.22561856e-03,
    ],
    [
        7.00000000e00,
        np.nan,
        np.nan,
        np.nan,
        1.90000000e-01,
        -1.50050000e-01,
        0.00000000e00,
        -1.90000000e-01,
        1.50050000e-01,
        0.00000000e00,
        np.nan,
        np.nan,
        np.nan,
    ],
    [
        0.00000000e00,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
    ],
    [
        0.00000000e00,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
        np.nan,
    ],
])


def createProblem():
    # Cannot use 1 model for the whole control cycle, because each model depends on the position of the feet
    # And the inertia matrix depends on the reference state (approximation )
    runningModels = []
    x0 = np.zeros(12)
    x0[2] = 0.2
    for i in range(N):
        model = ActionModelQuadruped()

        # Add model to the list of model
        runningModels.append(model)

    # Terminal Model
    terminalModel = ActionModelQuadruped()

    terminalModel.forceWeights = np.zeros((12, 1))
    terminalModel.frictionWeights = 0

    problem = crocoddyl.ShootingProblem(x0, runningModels, terminalModel)

    xs = [x0] * (N + 1)
    us = [np.zeros(12)] * N

    return xs, us, problem


def updateProblem(fsteps, xref, x0, problem):
    # Update the model according to the position of the feet in local frame (fstep)
    # Construction of the gait matrix representing the feet in contact with the ground
    index = next((idx for idx, val in np.ndenumerate(fsteps[:, 0]) if val == 0.0), 0.0)[0]
    gait[:, 0] = fsteps[:, 0]
    gait[:index, 1:] = 1.0 - (np.isnan(fsteps[:index, 1::3]) | (fsteps[:index, 1::3] == 0.0))

    # Replace NaN values by zeroes
    fsteps[np.isnan(fsteps)] = 0.0
    j = 0
    k_cum = 0
    # L = []

    # Iterate over all phases of the gait
    # The first column of xref correspond to the current state
    while gait[j, 0] != 0:
        for i in range(k_cum, k_cum + np.int(gait[j, 0])):

            # Update model
            problem.runningModels[i].updateModel(
                np.reshape(fsteps[j, 1:], (3, 4), order="F"),
                xref[:, i + 1],
                gait[j, 1:],
            )

        k_cum += np.int(gait[j, 0])
        j += 1

    # Update model of the terminal model
    problem.terminalModel.updateModel(np.reshape(fsteps[j - 1, 1:], (3, 4), order="F"), xref[:, -1], gait[j - 1, 1:])

    # update initial state of the problem
    problem.x0 = x0
    xs = [x0] * (N + 1)
    us = [np.zeros(12)] * N

    return xs, us, problem


def runUpdateProblemBenchmark(fsteps, xref, x0, problem):
    xs = [x0] * (N + 1)
    us = [np.zeros(12)] * N
    duration = []
    for i in range(T):
        c_start = time.time()
        xs, us, problem = updateProblem(fsteps, xref, x0, problem)
        c_end = time.time()
        duration.append(1e3 * (c_end - c_start))

    avrg_duration = sum(duration) / len(duration)
    min_duration = min(duration)
    max_duration = max(duration)
    return avrg_duration, min_duration, max_duration


def runDDPSolveBenchmark(xs, us, problem):
    ddp = crocoddyl.SolverDDP(problem)
    duration = []
    for i in range(T):
        c_start = time.time()
        ddp.solve(xs, us, MAXITER)
        c_end = time.time()
        duration.append(1e3 * (c_end - c_start))

    avrg_duration = sum(duration) / len(duration)
    min_duration = min(duration)
    max_duration = max(duration)
    return avrg_duration, min_duration, max_duration


def runShootingProblemCalcBenchmark(xs, us, problem):
    duration = []
    for i in range(T):
        c_start = time.time()
        problem.calc(xs, us)
        c_end = time.time()
        duration.append(1e3 * (c_end - c_start))

    avrg_duration = sum(duration) / len(duration)
    min_duration = min(duration)
    max_duration = max(duration)
    return avrg_duration, min_duration, max_duration


def runShootingProblemCalcDiffBenchmark(xs, us, problem):
    duration = []
    for i in range(T):
        c_start = time.time()
        problem.calcDiff(xs, us)
        c_end = time.time()
        duration.append(1e3 * (c_end - c_start))

    avrg_duration = sum(duration) / len(duration)
    min_duration = min(duration)
    max_duration = max(duration)
    return avrg_duration, min_duration, max_duration


print("Python bindings:")
xs, us, problem = createProblem()
xs, us, problem = updateProblem(fsteps, xref, x0, problem)
avrg_duration, min_duration, max_duration = runUpdateProblemBenchmark(fsteps, xref, x0, problem)
print("  UpdateModel [ms]: {0} ({1}, {2})".format(avrg_duration, min_duration, max_duration))
avrg_duration, min_duration, max_duration = runDDPSolveBenchmark(xs, us, problem)
print("  DDP.solve [ms]: {0} ({1}, {2})".format(avrg_duration, min_duration, max_duration))
avrg_duration, min_duration, max_duration = runShootingProblemCalcBenchmark(xs, us, problem)
print("  ShootingProblem.calc [ms]: {0} ({1}, {2})".format(avrg_duration, min_duration, max_duration))
avrg_duration, min_duration, max_duration = runShootingProblemCalcDiffBenchmark(xs, us, problem)
print("  ShootingProblem.calcDiff [ms]: {0} ({1}, {2})".format(avrg_duration, min_duration, max_duration))

File: benchmark/quadruped.cpp
///////////////////////////////////////////////////////////////////////////////
// BSD 3-Clause License
//
// Copyright (C) 2018-2019, LAAS-CNRS
// Copyright note valid unless otherwise stated in individual files.
// All rights reserved.
///////////////////////////////////////////////////////////////////////////////

#include "crocoddyl/core/actions/unicycle.hpp"
#include "crocoddyl/core/utils/callbacks.hpp"
#include "crocoddyl/core/solvers/ddp.hpp"
#include "crocoddyl/core/utils/timer.hpp"
#include <quadruped-walkgen/quadruped.hpp>

void updateModel(std::vector<boost::shared_ptr<quadruped_walkgen::ActionModelQuadruped> > running_models_2,
                 boost::shared_ptr<quadruped_walkgen::ActionModelQuadruped> terminal_model_2,
                 Eigen::Matrix<double, 6, 5> gait, Eigen::Matrix<double, 12, 17> xref,
                 Eigen::Matrix<double, 6, 13> fsteps, int N) {
  // Iterate over all the phases of the gait matrix
  // The first column of xref correspond to the current state = x0
  // Tmp is needed to use .data(), transformation of a column into a vector
  Eigen::Array<double, 1, 12> tmp = Eigen::Array<double, 1, 12>::Zero();
  int max_index = int(gait.block(0, 0, 6, 1).array().min(1.).matrix().sum());
  int k_cum = 0;

  for (int j = 0; j < max_index; j++) {
    for (int k = k_cum; k < k_cum + int(gait(j, 0)); ++k) {
      if (k < int(N)) {
        // Update model :
        tmp = fsteps.block(j, 1, 1, 12).array();
        running_models_2[k]->update_model(
            Eigen::Map<Eigen::Matrix<double, 3, 4> >(tmp.data(), 3, 4),
            Eigen::Map<Eigen::Matrix<double, 12, 1> >(xref.block(0, k + 1, 12, 1).data(), 12, 1),
            Eigen::Map<Eigen::Matrix<double, 4, 1> >(gait.block(j, 1, 1, 4).data(), 4, 1));
      }
    }
    k_cum += int(gait(j, 0));
  }

  tmp = fsteps.block(max_index - 1, 1, 1, 12).array();
  Eigen::Array<double, 1, 4> gait_tmp = Eigen::Array<double, 1, 4>::Zero();
  gait_tmp = gait.block(max_index - 1, 1, 1, 4).array();

  terminal_model_2->update_model(Eigen::Map<Eigen::Matrix<double, 3, 4> >(tmp.data(), 3, 4),
                                 Eigen::Map<Eigen::Matrix<double, 12, 1> >(xref.block(0, 16, 12, 1).data(), 12, 1),
                                 Eigen::Map<Eigen::Matrix<double, 4, 1> >(gait_tmp.data(), 4, 1));
  terminal_model_2->set_force_weights(Eigen::Matrix<double, 12, 1>::Zero());
  terminal_model_2->set_friction_weight(0);
}

int main(int argc, char* argv[]) {
  // The time of the cycle contol is 0.02s, and last 0.32s --> 16nodes
  // Control cycle during one gait period
  unsigned int N = 16;    // number of nodes
  unsigned int T = 1000;  // number of trials
  unsigned int MAXITER = 1;
  if (argc > 1) {
    T = atoi(argv[1]);
    MAXITER = atoi(argv[2]);
    ;
  }

  // Creating the initial state vector (size x12) [x,y,z,Roll,Pitch,Yaw,Vx,Vy,Vz,Wroll,Wpitch,Wyaw]
  // Perturbation of Vx = 0.2m.s-1
  Eigen::Matrix<double, 12, 1> x0;
  x0 << 0, 0, 0.2, 0, 0, 0, 0.2, 0, 0, 0, 0, 0;
  Eigen::Matrix<double, 4, 1> S;
  S << 1, 0, 0, 1;

  // Creating the reference state vector (size 12x16) to follow during the control cycle
  // Nullifying the Vx speed.
  Eigen::Matrix<double, 12, 1> xref_vector;
  xref_vector << 0, 0, 0.2, 0, 0, 0, 0, 0, 0, 0, 0, 0;
  Eigen::Matrix<double, 12, 17> xref;
  xref.block(0, 0, 12, 1) = x0;  // first vector is the initial state
  xref.block(0, 1, 12, 16) = xref_vector.replicate<1, 16>();

  // Creating the gait matrix : The number at the beginning represents the number of node spent in that position
  // 1 -> foot in contact with the ground :  0-> foot in the air
  Eigen::Matrix<double, 6, 5> gait;
  gait << 1, 1, 1, 1, 1, 7, 1, 0, 0, 1, 1, 1, 1, 1, 1, 7, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0;

  // Creating the fsteps matrix that represents the position of the feet during the whole control cycle (0.32s).
  // [nb , x1,y1,z1,  x2,y2,z2 ...] in local frame
  // The number at the beginning represents the number of node spent in that position
  // Here, the robot starts with 4 feet on the ground at the first node, then during 7 nodes (0.02s * 7)
  // The leg right front leg and left back leg are in the air ...etc

  Eigen::Matrix<double, 6, 13> fsteps;
  fsteps << 1, 0.19, 0.15, 0.0, 0.19, -0.15, 0.0, -0.19, 0.15, 0.0, -0.19, -0.15, 0.0, 7, 0.19, 0.15, 0.0, 0, 0, 0, 0,
      0, 0, -0.19, -0.15, 0.0, 1, 0.19, 0.15, 0.0, 0.19, -0.15, 0.0, -0.19, 0.15, 0.0, -0.19, -0.15, 0.0, 7, 0, 0, 0,
      0.19, -0.15, 0.0, -0.19, 0.15, 0.0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0;

  // Creating the Shoting problem that needs boost::shared_ptr<crocoddyl::ActionModelAbstract>

  // Cannot use 1 model for the whole control cycle, because each model depends on the position of the feet
  // And the inertia matrix depends on the reference state (approximation )
  std::vector<boost::shared_ptr<crocoddyl::ActionModelAbstract> > running_models;

  for (int i = 0; i < int(N); ++i) {  // 16 nodes
    boost::shared_ptr<crocoddyl::ActionModelAbstract> model =
        boost::make_shared<quadruped_walkgen::ActionModelQuadruped>();
    running_models.push_back(model);
  }

  boost::shared_ptr<crocoddyl::ActionModelAbstract> terminal_model;
  terminal_model = boost::make_shared<quadruped_walkgen::ActionModelQuadruped>();

  // Update each model and set to 0 the weight ont the command for the terminal node
  // For that, the internal method of quadruped_walkgen::ActionModelQuadruped needs to be accessed
  // -> Creation of a 2nd list using dynamic_cast

  int k_cum = 0;
  std::vector<boost::shared_ptr<quadruped_walkgen::ActionModelQuadruped> > running_models_2;

  // Iterate over all the phases of the gait matrix
  // The first column of xref correspond to the current state = x0
  // Tmp is needed to use .data(), transformation of a column into a vector
  Eigen::Array<double, 1, 12> tmp = Eigen::Array<double, 1, 12>::Zero();
  int max_index = int(gait.block(0, 0, 6, 1).array().min(1.).matrix().sum());

  for (int j = 0; j < max_index; j++) {
    for (int k = k_cum; k < k_cum + int(gait(j, 0)); ++k) {
      if (k < int(N)) {
        boost::shared_ptr<quadruped_walkgen::ActionModelQuadruped> model2 =
            boost::dynamic_pointer_cast<quadruped_walkgen::ActionModelQuadruped>(running_models[k]);
        running_models_2.push_back(model2);

        // Update model :
        tmp = fsteps.block(j, 1, 1, 12).array();
        model2->update_model(Eigen::Map<Eigen::Matrix<double, 3, 4> >(tmp.data(), 3, 4),
                             Eigen::Map<Eigen::Matrix<double, 12, 1> >(xref.block(0, k + 1, 12, 1).data(), 12, 1),
                             Eigen::Map<Eigen::Matrix<double, 4, 1> >(gait.block(j, 1, 1, 4).data(), 4, 1));
      }
    }
    k_cum += int(gait(j, 0));
  }

  boost::shared_ptr<quadruped_walkgen::ActionModelQuadruped> terminal_model_2 =
      boost::dynamic_pointer_cast<quadruped_walkgen::ActionModelQuadruped>(terminal_model);

  tmp = fsteps.block(max_index - 1, 1, 1, 12).array();
  Eigen::Array<double, 1, 4> gait_tmp = Eigen::Array<double, 1, 4>::Zero();
  gait_tmp = gait.block(max_index - 1, 1, 1, 4).array();

  terminal_model_2->update_model(Eigen::Map<Eigen::Matrix<double, 3, 4> >(tmp.data(), 3, 4),
                                 Eigen::Map<Eigen::Matrix<double, 12, 1> >(xref.block(0, 16, 12, 1).data(), 12, 1),
                                 Eigen::Map<Eigen::Matrix<double, 4, 1> >(gait_tmp.data(), 4, 1));
  terminal_model_2->set_force_weights(Eigen::Matrix<double, 12, 1>::Zero());
  terminal_model_2->set_friction_weight(0);

  boost::shared_ptr<crocoddyl::ShootingProblem> problem =
      boost::make_shared<crocoddyl::ShootingProblem>(x0, running_models, terminal_model);
  crocoddyl::SolverDDP ddp(problem);

  std::vector<Eigen::VectorXd> xs(int(N) + 1, x0);
  Eigen::Matrix<double, 12, 1> u0;
  u0 << 1, 0.2, 0.5, 1, 1, -0.2, -1, 1, 0.5, -1, -1, -0.5;
  std::vector<Eigen::VectorXd> us(int(N), u0);

  Eigen::ArrayXd duration(T);

  // Updating the problem
  for (unsigned int i = 0; i < T; ++i) {
    crocoddyl::Timer timer;
    updateModel(running_models_2, terminal_model_2, gait, xref, fsteps, N);
    duration[i] = timer.get_duration();
  }
  double avrg_duration = duration.sum() / T;
  double min_duration = duration.minCoeff();
  double max_duration = duration.maxCoeff();
  std::cout << "  UpdateModel [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Solving the optimal control problem
  for (unsigned int i = 0; i < T; ++i) {
    crocoddyl::Timer timer;
    ddp.solve(xs, us, MAXITER);
    duration[i] = timer.get_duration();
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  DDP.solve [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Solving the optimal control problem
  for (unsigned int i = 0; i < T; ++i) {
    crocoddyl::Timer timer;
    ddp.solve(xs, us, MAXITER);
    duration[i] = timer.get_duration();
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  DDP.solve [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Running calc
  for (unsigned int i = 0; i < T; ++i) {
    crocoddyl::Timer timer;
    problem->calc(xs, us);
    duration[i] = timer.get_duration();
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  ShootingProblem.calc [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Running calcDiff
  for (unsigned int i = 0; i < T; ++i) {
    crocoddyl::Timer timer;
    problem->calcDiff(xs, us);
    duration[i] = timer.get_duration();
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  ShootingProblem.calcDiff [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration
            << ")" << std::endl;
}